### ISSUE TITLE
An application cannot control SCTP level revoking

### ISSUE THREAD

**Issue by tuexen** [gloinul/draft-westerlund-tsvwg-dtls-over-sctp-bis#14]

**Comment by tuexen** [comment]
```
To prevent DTLS from discarding DTLS user messages while it is shutting down,
a CloseNotify message MUST only be sent after all outstanding SCTP user
messages have been acknowledged by the SCTP peer and MUST NOT be
revoked by the SCTP peer.¶
```
doesn't make sense. At least it needs rewording.

**Comment by gloinul** [comment]
That text was in RFC 6083.

To my understanding the issue here is that if the DTLS layer sends a protected user message  and immediately after that does SCTP association close, then if that close is acted on the DTLS record may not have been completely delivered. Which if partial delivered records are provided to the received implementation and then forces decryption and integrity verification, then the receiver will have DTLS connection failure also at the end. Thus, for a controlled close, each peer needs to first stop sending data, then verify that the data has been delivered on all streams before it does an close on the SCTP association. 

So yes, this do need an reformulation.

**Comment by tuexen** [comment]
Ahh, I see. Now I remember. That is the generic formulation for the sender dry dance.... The reason for this is that you need to avoid the DTLS layer to drop messages due to non-expected epoch values. So the sender sends everything with the old epoch, waits until the cum ack goes by (this is the revoking part) and then sends they new epoch. On the receiver side when getting the first first message for the new epoch, the receiver must empty the receive buffer first.
The point here is that only messages on stream 0 are in sequence...
I think this dance is still needed, but John thinks different. I'll open an issue for discussing this.

**Comment by gloinul** [comment]
I think this comes down to how strict the DTLS stack will be with epochs for individual records close to the change of keys. I think we really don't want epoch changes between DTLS records within a protected user message. So the sender would need to complete the protection operation of user messages being protected, then perform the key change, then it can process new messages. Which from a sender perspective introduce some delay but not horribly much. 

1. Indicated need to rekey, or renegotiation message received.
2. Lock for accepting new user messages
3. Empty sending queue, i.e. all data delivered to SCTP. 
4. Perform rekeying (may include message exchange with peer)
5. Unlock for new messages.

I think the receiver side is much harder if one can't retrieve the previous key for a short duration until one have ensured that each individual protected user message protected by the old key have been processed. We have to take into account that user messages on different streams may be down prioritized or subject to packet losses require retransmission.

**Comment by tuexen** [comment]
> I think this comes down to how strict the DTLS stack will be with epochs for individual records close to the change of keys. I think we really don't want epoch changes between DTLS records within a protected user message. So the sender would need to complete the protection operation of user messages being protected, then perform the key change, then it can process new messages. Which from a sender perspective introduce some delay but not horribly much.
> 
> 1. Indicated need to rekey, or renegotiation message received.
> 2. Lock for accepting new user messages
> 3. Empty sending queue, i.e. all data delivered to SCTP.
> 4. Perform rekeying (may include message exchange with peer)
> 5. Unlock for new messages.
> 
> I think the receiver side is much harder if one can't retrieve the previous key for a short duration until one have ensured that each individual protected user message protected by the old key have been processed. We have to take into account that user messages on different streams may be down prioritized or subject to packet losses require retransmission.

As I said, we need to drain the sender side (which is ensuring nothing is buffered anymore in the SCTP layer; when using the socket API, this is waiting for the sender dry event). Then send the new message. On the receiver side when getting this new message, buffer it, read all other messages from SCTP, and finally process it. This way we can make sure all oder messages are processed.

**Comment by gloinul** [comment]
Okay, I think we need to investigate if we have two modes of operation here. We have the guaranteed to work, which is to drain the queue, but can cause impact on the upper layer protocol as we might introduce some delay when messages are queue while performing the drain operation. The second one is to know that the peer supports processing the previous epoch for a while, so that one can just switch key on the next user message one starts sending. 

Both of these would still have issues issues with very long user messages. So I think we need more thinking on this issues.

**Comment by gloinul** [comment]
Issue #25 is related to this discussion.

**Comment by gloinul** [comment]
The text needs reformulation. @tuexen will reword this to be clearer about the need for doing draining before closing the SCTP association.
